File: clients/scala/src/main/scala/com/priostack/acn/Exceptions.scala
package com.priostack.acn

/** Typed errors for the Priostack ACN client.
  *
  * The ACN reports two very different kinds of failure and the client raises a
  * different branch of this hierarchy for each:
  *
  *   - '''Transport / protocol failures''' — the request never produced a valid
  *     tool envelope: a network error, an HTTP error status, malformed JSON, or
  *     a JSON-RPC `error` object (unknown method, bad params). These raise
  *     [[AcnTransportError]].
  *   - '''Tool failures''' — the call reached the tool but the tool declined it:
  *     the envelope came back with `ok: false` and an `outcome` such as
  *     `capability-denied` or `not-found`. These raise the matching subclass of
  *     [[AcnToolError]], so callers can catch exactly the outcome they expect.
  *
  * Every error derives from [[AcnError]], so a single `catch { case _: AcnError }`
  * covers everything the client can raise.
  */
abstract class AcnError(
    message: String,
    val method: Option[String] = None,
    cause: Throwable = null
) extends RuntimeException(message, cause)

/** A network, HTTP, decoding, or JSON-RPC protocol failure.
  *
  * Raised before a valid tool envelope could be obtained: connection errors,
  * timeouts, non-2xx HTTP responses, non-JSON bodies, envelopes missing the
  * `content[0].text` payload, or a JSON-RPC `error` object.
  */
final class AcnTransportError(
    message: String,
    method: Option[String] = None,
    val code: Option[Int] = None,
    val httpStatus: Option[Int] = None,
    cause: Throwable = null
) extends AcnError(message, method, cause)

/** A tool declined the call (envelope `ok: false`).
  *
  * Carries the server `outcome` string and human-readable `detail`. Prefer
  * catching a specific subclass; catch this base only when any tool-level
  * failure should be handled the same way. Unknown outcomes (e.g. a new server
  * outcome) surface as this base class rather than being mis-mapped.
  */
class AcnToolError(
    val detail: String,
    val outcome: String,
    method: Option[String] = None,
    val data: Option[ujson.Value] = None
) extends AcnError(
      if (detail.nonEmpty) detail else if (outcome.nonEmpty) outcome else "ACN tool error",
      method
    )

/** `not-found` — no such session, space, or entity. */
final class NotFoundError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "not-found", method, data)

/** `invalid-query` — malformed arguments or an unknown enum value. */
final class InvalidQueryError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "invalid-query", method, data)

/** `capability-denied` — missing capability, or an invalid/revoked token. */
final class CapabilityDeniedError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "capability-denied", method, data)

/** `policy-denied` — a governance policy refused the operation. */
final class PolicyDeniedError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "policy-denied", method, data)

/** `requires-governance` — the change needs an explicit confirm step. */
final class RequiresGovernanceError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "requires-governance", method, data)

/** `stale-base` — the operation raced a newer state; re-read and retry. */
final class StaleBaseError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "stale-base", method, data)

/** `integrity-fault` — an internal consistency check failed. */
final class IntegrityFaultError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "integrity-fault", method, data)

/** `conflict` — the write conflicts with existing state. */
final class ConflictError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "conflict", method, data)

/** `capacity-exhausted` — a tier/registration/store ceiling was hit. */
final class CapacityExhaustedError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "capacity-exhausted", method, data)

/** `not-implemented` — the tool is unavailable on this ACN deployment.
  *
  * Most commonly: calling [[AcnClient.register]] against a single-tenant/offline
  * ACN that does not offer self-registration.
  */
final class NotSupportedError(detail: String, method: Option[String] = None, data: Option[ujson.Value] = None)
    extends AcnToolError(detail, "not-implemented", method, data)

object AcnToolError:
  /** Build the most specific [[AcnToolError]] for a server `outcome`. Unknown
    * outcomes fall back to the [[AcnToolError]] base so a new server outcome
    * degrades gracefully instead of being mis-mapped.
    */
  def apply(
      outcome: String,
      detail: String,
      method: Option[String] = None,
      data: Option[ujson.Value] = None
  ): AcnToolError =
    outcome match
      case "not-found"           => new NotFoundError(detail, method, data)
      case "invalid-query"       => new InvalidQueryError(detail, method, data)
      case "capability-denied"   => new CapabilityDeniedError(detail, method, data)
      case "policy-denied"       => new PolicyDeniedError(detail, method, data)
      case "requires-governance" => new RequiresGovernanceError(detail, method, data)
      case "stale-base"          => new StaleBaseError(detail, method, data)
      case "integrity-fault"     => new IntegrityFaultError(detail, method, data)
      case "conflict"            => new ConflictError(detail, method, data)
      case "capacity-exhausted"  => new CapacityExhaustedError(detail, method, data)
      case "not-implemented"     => new NotSupportedError(detail, method, data)
      case _                     => new AcnToolError(detail, outcome, method, data)

File: clients/kotlin/src/main/kotlin/com/priostack/acn/Exceptions.kt
/*
 * Typed exceptions for the Priostack ACN client.
 *
 * The ACN reports two very different kinds of failure and the client raises a
 * different branch of this hierarchy for each:
 *
 *  * Transport / protocol failures — the request never produced a valid tool
 *    envelope: a network error, an HTTP error status, malformed JSON, or a
 *    JSON-RPC `error` object (unknown method, bad params). These raise
 *    [AcnTransportError].
 *
 *  * Tool failures — the call reached the tool but the tool declined it: the
 *    envelope came back with `ok: false` and an `outcome` such as
 *    `capability-denied` or `not-found`. These raise the matching subclass of
 *    [AcnToolError], so callers can catch exactly the outcome they expect.
 *
 * Every exception derives from the sealed [AcnException], so a single
 * `catch (e: AcnException)` handles anything this client can raise.
 */
package com.priostack.acn

import kotlinx.serialization.json.JsonElement

/** Base class for every error raised by [AcnClient]. */
sealed class AcnException(
    message: String,
    /** The tool name (`noetic.*`) the failing call targeted, if known. */
    val method: String? = null,
    cause: Throwable? = null,
) : RuntimeException(message, cause)

/**
 * A network, HTTP, decoding, or JSON-RPC protocol failure.
 *
 * Raised before a valid tool envelope could be obtained: connection errors,
 * timeouts, non-2xx HTTP responses, non-JSON bodies, envelopes missing the
 * `content[0].text` payload, or a JSON-RPC `error` object.
 */
class AcnTransportError(
    message: String,
    method: String? = null,
    /** JSON-RPC error code, when the failure was a JSON-RPC `error`. */
    val code: Int? = null,
    /** HTTP status code, when the failure was an HTTP error. */
    val httpStatus: Int? = null,
    cause: Throwable? = null,
) : AcnException(message, method, cause)

/**
 * A tool declined the call (envelope `ok: false`).
 *
 * Carries the server [outcome] string and human-readable [detail]. Prefer
 * catching a specific subclass; catch this base only when any tool-level
 * failure should be handled the same way. An unrecognised outcome degrades to
 * this base class rather than being mis-mapped.
 */
open class AcnToolError internal constructor(
    /** The server outcome, e.g. `"capability-denied"`. */
    val outcome: String,
    /** The server's `detail` message, if any. */
    val detail: String,
    method: String? = null,
    /** Any partial `data` the server attached to the failure. */
    val data: JsonElement? = null,
) : AcnException(detail.ifEmpty { outcome.ifEmpty { "ACN tool error" } }, method)

/** `not-found` — no such session, space, or entity. */
class NotFoundError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("not-found", detail, method, data)

/** `invalid-query` — malformed arguments or an unknown enum value. */
class InvalidQueryError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("invalid-query", detail, method, data)

/** `capability-denied` — missing capability, or an invalid/revoked token. */
class CapabilityDeniedError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("capability-denied", detail, method, data)

/** `policy-denied` — a governance policy refused the operation. */
class PolicyDeniedError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("policy-denied", detail, method, data)

/** `requires-governance` — the change needs an explicit confirm step. */
class RequiresGovernanceError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("requires-governance", detail, method, data)

/** `stale-base` — the operation raced a newer state; re-read and retry. */
class StaleBaseError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("stale-base", detail, method, data)

/** `integrity-fault` — an internal consistency check failed. */
class IntegrityFaultError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("integrity-fault", detail, method, data)

/** `conflict` — the write conflicts with existing state. */
class ConflictError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("conflict", detail, method, data)

/** `capacity-exhausted` — a tier/registration/store ceiling was hit. */
class CapacityExhaustedError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("capacity-exhausted", detail, method, data)

/**
 * `not-implemented` — the tool is unavailable on this ACN deployment.
 *
 * Most commonly: calling [AcnClient.register] against a single-tenant/offline
 * ACN that does not offer self-registration.
 */
class NotSupportedError internal constructor(detail: String, method: String? = null, data: JsonElement? = null) :
    AcnToolError("not-implemented", detail, method, data)

/**
 * Build the most specific [AcnToolError] for a server `outcome`. Unknown
 * outcomes fall back to the [AcnToolError] base so a new server outcome
 * degrades gracefully instead of being mis-mapped.
 */
internal fun toolErrorFor(
    outcome: String,
    detail: String,
    method: String? = null,
    data: JsonElement? = null,
): AcnToolError = when (outcome) {
    "not-found" -> NotFoundError(detail, method, data)
    "invalid-query" -> InvalidQueryError(detail, method, data)
    "capability-denied" -> CapabilityDeniedError(detail, method, data)
    "policy-denied" -> PolicyDeniedError(detail, method, data)
    "requires-governance" -> RequiresGovernanceError(detail, method, data)
    "stale-base" -> StaleBaseError(detail, method, data)
    "integrity-fault" -> IntegrityFaultError(detail, method, data)
    "conflict" -> ConflictError(detail, method, data)
    "capacity-exhausted" -> CapacityExhaustedError(detail, method, data)
    "not-implemented" -> NotSupportedError(detail, method, data)
    else -> AcnToolError(outcome, detail, method, data)
}
